WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,260 @@
+                return await withCheckedContinuation { continuation in
+                    orderPaymentUseCase.collectPayment(using: discoveryMethod) { error in
+                        // TODO: even though we have a tri-state result type, perhaps we should throw these errors.
+                        if let error = error as? CardPaymentErrorProtocol {
+                            continuation.resume(returning: CardPresentPaymentResult.failure(error))
+                        } else {
+                            continuation.resume(returning: CardPresentPaymentResult.failure(CardPaymentsAdaptorError.unknownPaymentError(underlyingError: error)))
+                        }
+                    } onCancel: {
+                        continuation.resume(returning: CardPresentPaymentResult.cancellation)
+                    } onPaymentCompletion: {
+                        // no-op – not used in PaymentMethodsViewModel anyway so this can be removed
+                    } onCompleted: {
+                        continuation.resume(returning: CardPresentPaymentResult.success(order))
+                    }
+                }
+            } onCancel: {
+                // TODO: cancel any in-progress discovery, connection, or payment.
+            }
+        }
+        self.paymentTask = paymentTask
+
+        return await paymentTask.value
+    }
+
+    func cancelPayment() {
+        paymentTask?.cancel()
+        paymentScreenEventSubject.send(nil) // This removes any otherwise-presented UI
+    }
+
+    enum CardPaymentsAdaptorError: Error, CardPaymentErrorProtocol {
+        var retryApproach: CardPaymentRetryApproach {
+            .restart
+        }
+
+        case unknownPaymentError(underlyingError: Error)
+    }
+}
+
+extension CardPresentPaymentsAdaptor: CardPresentPaymentAlertsPresenting {
+
+    func present(viewModel: CardPresentPaymentsModalViewModel) {
+        paymentScreenEventSubject.send(.presentAlert(viewModel as CardPresentPaymentsModalContent))
+    }
+
+    func foundSeveralReaders(readerIDs: [String], connect: @escaping (String) -> Void, cancelSearch: @escaping () -> Void) {
+        paymentScreenEventSubject.send(.presentReaderList(readerIDs))
+        // the button actions here might need to be communicated... or we could expose them on the adaptor somehow.
+    }
+
+    func updateSeveralReadersList(readerIDs: [String]) {
+        paymentScreenEventSubject.send(.presentReaderList(readerIDs))
+    }
+
+    func dismiss() {
+        paymentScreenEventSubject.send(nil)
+        // TODO: Decide whether we really need this
+    }
+}
+
+import Combine
+
+/// This is really a re-implementation of the CardPresentPaymentsOnboardingPresenter, as it needs to take the calls to `showOnboardingIfRequired` and
+/// route the output to a SwiftUI view for display, rather than directly displaying on the viewController that's passed in.
+final class CardPresentPaymentsOnboardingPresenterAdaptor: CardPresentPaymentsOnboardingPresenting {
```

Comment:
From what I understand this adapter is just a copy of the existing `CardPresentPaymentsOnboardingPresenter` to be used through our SwiftUI implementation, it just ignores the usage of the `UIViewController` property within `showOnboardingIfRequired(:)`, that we need to declare to conform to the protocol, right?

Reply:
Yep pretty much. 

It _replaces_ the use of that UIViewController property with use of its `onboardingScreenViewModelPublisher` to emit the UI events for our new code to display.

In the older implementation, the `CardPresentPaymentsOnboardingPresenter` has the responsibility for actually presenting the UI, which this adaptor doesn't.